scripts/docker/build_docker.py

```diff
@@ -107,76 +247,118 @@ def go_to_workdir(self):
+                print(f"changed_project_files: {changed_project_files}", file=sys.stderr)
+                self.project_arguments.targets = [
+                    target for target, image_dependencies in self.dependencies.items()
+                    if image_dependencies.has_change(changed_project_files)
+                ]
+                print(f"targets = {self.project_arguments.targets}")
+            elif "all" in self.project_arguments.targets:
+                self.project_arguments.targets = ProjectBuilder.images_built_by_all
+            # get targets + their dependencies in order (so that builds succeed)
+            expanded_build_targets = self.get_ordered_build_chain_list()
+
+            # build each required dependency
+            print("Building and pushing the following targets in order:" if self.remote_docker_repos else
+                  "Building the following targets in order:")
+            print(expanded_build_targets)
+            if self.project_arguments.dry_run:
+                for target_name in expanded_build_targets:
+                    ImageBuilder(target_name, self).update_current_image()
+            else:
+                print(colored('#################################################', 'magenta'))
+                for target_name in expanded_build_targets:
+                    a = colored("Building image ", "grey")
+                    b = colored(target_name + ":" + self.project_arguments.image_tag, "yellow", attrs=['bold'])
+                    c = colored(" ...", "grey")
+                    print(a, b, c)
+
+                    build_time_args = {
+                        arg: self.get_current_image(image_name)
+                        for image_name, arg in ProjectBuilder.dependencies[target_name].docker_dependencies.items()
+                    }
+
+                    image_builder = ImageBuilder(target_name, self)
+                    image_builder.build(build_time_args)
+                    image_builder.push()
+                    print(colored('#################################################', 'magenta'))
+
+                print(colored('BUILD PROCESS SUCCESS!', 'green'))
+
+            if self.remote_docker_repos:
+                # push any images that are purely local (this can happen if e.g. images are built without specifying
+                # --docker-repo during development, but upon successful build the developer wants to push them)
+                local_images_to_push = [
+                    image
+                    for image in [
+                        self.get_current_image(target) for target in self.dependencies.keys()
+                    ]
+                    if ProjectBuilder.image_is_local(image)
+                ]
+                if local_images_to_push:
+                    print(f"Also push local images {local_images_to_push} to {self.remote_docker_repos}")
+                if not self.project_arguments.dry_run:
```

Comment:
I'm not following the objection here. Dry-run doesn't update dockers.json, but does tell you how it *would* be changed, so it lets you know what would happen. It's possible to run the script without pushing to a remote repo (maybe you want to debug some stuff first without pushing?). It's also set up to allow you to push local images built on a previous run to remote images. And dry-run does tell you what it would do in each of those cases.

Reply:
> but does tell you how it would be changed

Reply:
The high-level view of how the program works is:

1. Target selection. This is either manual (via `--targets`) or automatic (via changes detected from git)
2. Determine which images must be built. Done by finding images that depend (directly or indirectly on targets). There is the `--no-force-rebuild` option that doesn't build an image or its dependencies if an image with the desired tag already exists (presumably useful as part of a debugging process where several Dockerfiles are being tweaked).
3. Actually build the images that must be built.
4. Push the newly built images. If there is a remote image repo specified, push any newly built images to that repo and update dockers.json with the remote image. If there are pre-existing local images in dockers.json that were not built this run, push them now and update dockers.json with their remote name. If there is no remote image repo, update dockers.json with the local image.

When you run the program in normal mode, it prints statements describing its actions at each step, including printing out changes to dockers.json. When you run it with `--dry-run` it does all the same stuff, except it never builds and image, never pushes and image and doesn't actually alter dockers.json, instead it prints the changes that it _would_ make if you ran without `--dry-run`. It has the full list of images that would be built from step 2, and it can see the local images by loading in the input dockers.json, so actually building and pushing are unnecessary to determing how dockers.json would be changed.

I added more verbosity for the descriptions of the `--docker-repo` and `--input-json` arguments to hopefully make it easier to understand how to use the program's full functionality.

Reply:
Done. I added it below the imports since pycharm's linter complained otherwise.